Block Comments: Improve input labels

karthick-murugan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes #71681

Improves comment input labels for better accessibility by implementing sequential numbering and visible labels for comment forms.


<!-- In a few words, what is the PR actually doing? -->

## Why?
Currently, all comment inputs have the same generic "Comment" label, which provides no context for screen reader users. This makes the collaborative commenting feature largely unusable for users with assistive technologies, as they cannot distinguish between different comment contexts or know which comment they're replying to.

This PR addresses the accessibility issue by implementing unique, contextual labels for each comment input, following the recommendations from @joedolson in the issue discussion.

## How

1. **Sequential Comment Numbering**: Comments are now numbered sequentially (Comment 1, Comment 2, etc.) based on creation order, not content order.

2. **Contextual Labels**: 
   - New comments: "Comment X" (where X is the next sequential number)
   - Reply forms: "Reply to Comment X" 
   - Edit forms: "Edit Comment X"

3. **Visible Labels**: Labels are now visible to all users (not just screen readers) to facilitate team communication.

4. **Smart Label Placement**: Comment labels appear in the header area above the gravatar image for better visual hierarchy.

### Implementation Details:

- Enhanced `CommentForm` component with `labelText` props
- Updated `CommentBoard` component with `showLabel` prop to control label visibility
- Modified `Comments` component to pass sequential numbering to child components
- Updated `AddComment` component to calculate next comment number
- Added proper styling for visible labels

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->

1. Open a post or page in the Gutenberg editor
2. Select a block and add a comment using the collaboration sidebar
3. Verify that the comment input shows "Comment 1" label
4. Add another comment to a different block
5. Verify that the second comment input shows "Comment 2" label
6. Click on the first comment to expand replies
7. Verify that the reply input shows "Reply to Comment 1" label
8. Add a reply and verify it appears without any label
9. Test editing a comment and verify it shows "Edit Comment X" label


### Video 


https://github.com/user-attachments/assets/7437c8cc-dbd7-454c-b658-13987f66afed

karthick-murugan committed:
Improve comment input labels

karthick-murugan committed:
Removed return to block function

t-hamano commented:
I'm not sure if using sequential numbers for labels is the right approach. For example, someone asks someone else to review comment 3. Then someone else deletes comment 2. Now comment 3 has become comment 2. If we add a number, I think it should be a comment ID.

Personally, I think the visual label is too obtrusive. My suggestion is to make the comment container "truly" focusable and give it a more specific label.

For example, in Google Docs, the comment container is a list item and is focusable. When the container is focused, a screen reader will announce: 


```
Comments dialog. Open comment. Author Aki Hamano. 1 replies. 0 new replies.  1 of 2  level 1
```

cc @joedolson

karthick-murugan commented:
@t-hamano - Thank you for the detailed feedback! I've implemented the changes in my local as you suggested and tested them with VoiceOver. Still not yet pushed the code. Here's what I've addressed:

**Changes Made**

**Removed Visual Labels:** Eliminated the obtrusive visual labels as you suggested - the interface is now cleaner and less cluttered.
**Enhanced Accessibility:** Implemented rich, contextual aria-labels that provide comprehensive information to screen readers, including:
Comment context ("Comments dialog. Open comment")
Author information
Reply count
Status (active/resolved)
Posting date
Clear interaction instructions
**Focusable Containers:** Made comment containers truly focusable with proper keyboard navigation support.

**Question About List Items**

Regarding your mention of list items - could you clarify what you mean by "the contents of the comments container list items"?
Are you suggesting that each comment should be wrapped in `<li>` elements within a `<ul>`? Something else regarding the list structure?

Here is the video of the tested implementation. Please have a look and let me know your feedback.

https://github.com/user-attachments/assets/74708331-d12a-4967-9665-2d71fde274bb

t-hamano commented:
@karthick-murugan Sorry, my suggestions should be addressed in a separate PR. I have submitted #71883 to make threads focusable and improve keyboard navigation. In this PR (#71883), we should be able to simply focus on improving the labels.

> I think sequential is fine; but the numbers should be persistent. If you delete comment 2, that should not change the labels on any other comment.

I'm not sure how to make the numbers persistent other than using the comment ID. Any ideas? The current PR generates sequential numbers based on the length of the thread, so the comment numbers could change.

jasmussen commented:
Thanks for the updates.

> Clicking on comments should scroll to the associated block and highlight it. The spotlight mode is a global setting, so I would avoid enabling it on comment click.

My mistake, I was using an old branch. Here's how that looks:

![focus](https://github.com/user-attachments/assets/a900223b-45f1-4f5d-9779-1e1809a1b9aa)

I still feel like we need a toolbar change. The previously selected block has a toolbar that's sticky to the top of the screen.

> The spotlight mode is a global setting, so I would avoid enabling it on comment click.

I want to elaborate more on this, because just to be clear I'm not suggesting the user preference is changed at all. If you already have it on, you'll see no change. If you already have it off, the check in the dropdown will not be added, it's only the _effect_ that will be applied, and only when the comment itself is selected or has focus inside. The idea is to use the scaffolding already built for spotlight mode, but apply it in more places. The feature itself is very underused right now.

To visualize it I created [this prototype](https://www.figma.com/proto/CSlNU8T0EQ0dhLVIppNwFl/Annotations-Comments-Collaborative-Editing?page-id=629%3A17351&node-id=629-17730&p=f&viewport=46%2C609%2C0.45&t=TbmRHj2FVavztslL-1&scaling=min-zoom&content-scaling=fixed&starting-point-node-id=629%3A17730):

https://github.com/user-attachments/assets/a8c11fdf-48a1-4f79-a7df-b69c0b49da31

As noted, I don't think it's a blocker, but let me know if the behavior makes sense!

Mamaduka commented:
@jasmussen, I think I understand it better. We should probably create a separate issue for this and move the discussion there.

karthick-murugan committed:
Feedback Changes

karthick-murugan committed:
Merge branch 'trunk' into improve-comment-input-labels

karthick-murugan commented:
@Mamaduka @t-hamano 

Updated the conflicts and I have updated the labels such like "Comments dialog: Open comment 196. Author admin 0 replies. Status: active. Posted 22/09/2025" where 196 is the commend id. This was implemented based on our earlier discussions. Please review the video, and let me know if further adjustments are needed.

https://github.com/user-attachments/assets/ffc711f1-e982-471d-9c64-eb8476599968

karthick-murugan committed:
Removed few unwanted periods

t-hamano commented:
> P.S. Why did we decide to use "scroll to + highlight" when a comment is selected, instead of selecting the block? The latter seems to be the most common UX in other apps.

selectBlock moves focus to the block, which I don't think is ideal from an a11y perspective, so we went with the scroll+highlight approach. See: https://github.com/WordPress/gutenberg/pull/71308#pullrequestreview-3183358005

Mamaduka commented:
> I think the purpose of this PR is not to make threads focusable, but to only improve the label on the textarea itself in the comment form to make it clearer. What do you think?

100%. Let's only focus only on that. Atomic PRs/commits are always easy to review and merge.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -88,6 +88,7 @@ export function Comments( {
 		<Thread
 			key={ thread.id }
 			thread={ thread }
+			commentId={ thread.id }
```
This value is already available via the `thread` object. There's no need to pass additional props.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -239,6 +280,7 @@ function Thread( {
 
 const CommentBoard = ( {
 	thread,
+	commentId,
```
Same here.

karthick-murugan replied:
@Mamaduka - Updated both the feedbacks. Thank You.

karthick-murugan committed:
Updated the feedback

t-hamano commented on packages/editor/src/components/collab-sidebar/add-comment.js:
```diff
@@ -53,6 +53,10 @@ export function AddComment( {
 
+	// Generate contextual label for new comment
+	// For new comments, use a generic label since we don't have an ID yet
+	const commentLabel = _x( 'New Comment', 'Add new comment label' );
```
```suggestion
	const commentLabel = __( 'New Comment' );
```

The code comments and context seem redundant to me.

t-hamano commented on packages/editor/src/components/collab-sidebar/comment-form.js:
```diff
@@ -47,10 +49,13 @@ function CommentForm( {
-				{ __( 'Comment' ) }
-			</VisuallyHidden>
+		<VStack spacing="3">
+			<label
+				htmlFor={ inputId }
+				className="editor-collab-sidebar-panel__comment-label"
+			>
+				{ labelText || __( 'Comment' ) }
+			</label>
```
Instead of using a label element, I think it would be better to reuse the VisuallyHidden component.

Mamaduka replied:
+1.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -139,6 +139,31 @@ function Thread( {
 			id={ thread.id }
 			spacing="3"
 			onClick={ () => handleCommentSelect( thread ) }
+			onKeyDown={ ( event ) => {
+				if ( event.key === 'Enter' || event.key === ' ' ) {
+					event.preventDefault();
+					handleCommentSelect( thread.id );
+				}
+			} }
+			role="button"
+			tabIndex={ 0 }
+			aria-label={ sprintf(
+				// translators: %1$s: comment identifier, %2$s: author name, %3$s: reply count, %4$s: status, %5$s: date.
+				_x(
+					'Comments dialog. Open comment %1$s. Author %2$s. %3$s replies. Status: %4$s. Posted %5$s. Click to focus.',
+					'Comment accessibility label with full context'
+				),
+				thread.id,
+				thread?.author_name || 'Unknown',
+				thread?.replies?.length || 0,
+				thread?.status === 'approved' ? 'resolved' : 'active',
+				thread?.date
+					? new Date( thread.date ).toLocaleDateString()
+					: 'unknown date'
+			) }
+			{ ...( thread?.replies?.length > 0 && {
+				'aria-expanded': isFocused,
+			} ) }
```
I don't think these changes are necessary; the aria-label on the thread container itself could be discussed in https://github.com/WordPress/gutenberg/pull/71883.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -229,6 +254,15 @@ function Thread( {
 							}
 							rows={ 'approved' === thread.status ? 2 : 4 }
+							labelText={ sprintf(
+								// translators: %1$s: comment identifier, %2$s: author name
+								_x(
+									'Reply to Comment %1$s by %2$s',
+									'Reply to specific comment with author context'
+								),
```
```suggestion
								// translators: %1$s: comment identifier, %2$s: author name
								__( 'Reply to Comment %1$s by %2$s' ),
```

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -354,6 +388,15 @@ const CommentBoard = ( {
 					thread={ thread }
 					submitButtonText={ _x( 'Update', 'verb' ) }
+					labelText={ sprintf(
+						// translators: %1$s: comment identifier, %2$s: author name.
+						_x(
+							'Edit Comment %1$s by %2$s',
+							'Edit specific comment with author context'
+						),
```
```suggestion
						// translators: %1$s: comment identifier, %2$s: author name.
						__( 'Edit Comment %1$s by %2$s' ),
```

karthick-murugan replied:
@t-hamano - Updated all the changes.

karthick-murugan committed:
JS Lint fixes

joedolson commented:
While I generally like visible labels, I think it's OK for these not to be visible. These are single-field forms, triggered using a button that says 'Comment', and with a submit button that visually says 'Comment' or 'Reply'. With good visual indication of which block a comment is associated with, there isn't a lot that a visible label on the field will add, and it doesn't need to be distinguished from other nearby input fields. It's very similar to a search input in that respect.

joedolson reviewed (approved):
LGTM!

Mamaduka commented on packages/editor/src/components/collab-sidebar/style.scss:
```diff
@@ -43,6 +43,42 @@
+		}
+	}
+}
+
```
These classes aren't used anymore.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comment-form.js:
```diff
@@ -49,13 +47,10 @@ function CommentForm( {
-				{ labelText || __( 'Comment' ) }
-			</label>
+		<>
+			<VisuallyHidden as="label" htmlFor={ inputId }>
+				{ __( 'Comment' ) }
+			</VisuallyHidden>
```
Other components are still passing the label text, but it appears to have been removed from here, likely by accident.

```js
{ labelText ?? __( 'Comment' ) }
```

karthick-murugan replied:
@Mamaduka - Updated the changes

karthick-murugan committed:
Feedback changes

Mamaduka commented:
Thanks for your effort, @karthick-murugan!

Do you mind rebasing your branch on top of the latest trunk? We merged the end-to-end test coverage yesterday and want to ensure that we update locators as needed.

karthick-murugan committed:
Merge branch 'trunk' of github.com:karthick-murugan/gutenberg into improve-comment-input-labels

karthick-murugan commented:
@Mamaduka - Rebased my branch with trunk. Thank You.

karthick-murugan committed:
Merge branch 'trunk' into improve-comment-input-labels

karthick-murugan committed:
Js Lint errors
